streaming/aggregate/avg.go

```diff
@@ -0,0 +1,109 @@
+package aggregate
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type Average struct {
+	underlyingSum   *Sum
+	underlyingCount *Count
+}
+
+func NewAverageAggregate() *Average {
+	return &Average{
+		underlyingSum:   NewSumAggregate(),
+		underlyingCount: NewCountAggregate(),
+	}
+}
+
+func (agg *Average) AddValue(ctx context.Context, tx storage.StateTransaction, value octosql.Value) error {
+	valueType := value.GetType()
+
+	if !isAppropriateType(valueType) {
+		return errors.Errorf("type of value passed (%s) isn't appropriate for calculating average", valueType)
+	}
+
+	err := agg.underlyingSum.AddValue(ctx, tx, value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't add value to sum for avg")
+	}
+
+	err = agg.underlyingCount.AddValue(ctx, tx, value)
+	if err != nil {
+		err2 := agg.underlyingSum.RetractValue(ctx, tx, value) // TODO - is this necessary
+		if err2 != nil {
+			return errors.Wrap(err2, "couldn't retract value from sum for avg after error occured in count update")
+		}
+
+		return errors.Wrap(err, "couldn't add element to elements count for avg")
+	}
+
+	return nil
+}
+
+func isAppropriateType(valueType octosql.Type) bool {
+	return valueType == octosql.TypeInt || valueType == octosql.TypeFloat
+}
+
+func (agg *Average) RetractValue(ctx context.Context, tx storage.StateTransaction, value octosql.Value) error {
+	valueType := value.GetType()
+
+	if !isAppropriateType(valueType) {
+		return errors.Errorf("type of value passed (%s) isn't appropriate for calculating average", valueType)
+	}
+
+	err := agg.underlyingSum.RetractValue(ctx, tx, value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't retract value from sum for avg")
+	}
+
+	err = agg.underlyingCount.RetractValue(ctx, tx, value)
+	if err != nil {
+		err2 := agg.underlyingSum.AddValue(ctx, tx, value) // TODO - is this necessary
+		if err2 != nil {
+			return errors.Wrap(err2, "couldn't add value to sum for avg after error occured in count update")
+		}
+
+		return errors.Wrap(err, "couldn't add element to elements count for avg")
+	}
+
+	return nil
+}
+
+func (agg *Average) GetValue(ctx context.Context, tx storage.StateTransaction) (octosql.Value, error) {
+	currentSum, err := agg.underlyingSum.GetValue(ctx, tx)
+	if err != nil {
+		return octosql.ZeroValue(), errors.Wrap(err, "couldn't get current sum for avg")
+	}
+
+	currentCount, err := agg.underlyingCount.GetValue(ctx, tx)
+	if err != nil {
+		return octosql.ZeroValue(), errors.Wrap(err, "couldn't get current elements count for avg")
+	}
+
+	if currentCount.AsInt() <= 0 { // no values or early retractions
+		return octosql.MakeInt(0), nil // TODO - I don't think returning anything other than "neutral element" would be better
+	}
+
+	var currentAvg octosql.Value
+
+	switch currentSum.GetType() {
+	case octosql.TypeInt:
+		currentAvg = octosql.MakeFloat(float64(currentSum.AsInt()) / float64(currentCount.AsInt()))
```

Comment:
Is there a specific reason you do float64(currentSum.AsInt()) instead of currentSum.AsFloat() ? (same goes for currentCount)

Reply:
yea good point, that was needed before refactor, cause instead of Sum + Count there was Avg + Count (will fix this)
currentCount is always of TypeInt, so we have to typecast it

Reply:
some time ago i wrote that MakeInt(n).AsFloat() actually returns 0.0 for every positive n :P which I actually discovered in this exact place

Reply:
I'm sorry I misunderstood your comment. I need to typecast both currentSum and currentCount to floats in order to calculate currentAvg correctly. Therefore, this piece of code is truly needed.
@JasiekChomiak